# param/ndt_localizer.param.yaml
---
/**:
  ros__parameters:
    # Mapper specific configuration:
    file_name_prefix: "ndt_sample_map"
    publish_map_increment: true
    map:
      capacity: 1000000
      min_point:
        x: -1000.0
        y: -1000.0
        z: -3.0
      max_point:
        x: 1000.0
        y: 1000.0
        z: 3.0
      voxel_size:
        x: 1.0
        y: 1.0
        z: 1.0
      frame_id: map
    map_increment_pub:  # Config of the input point cloud subscription
      history_depth: 10
    ##### Relative localization node configuration:
    observation_sub:  # Config of the input point cloud subscription
      history_depth: 10
    # Config of the maps point clouds to register
    pose_pub:
      history_depth: 10
    # Publish the result to `/tf` topic
    publish_tf: true
    # Maximum allowed difference between the initial guess and the ndt pose estimate
    predict_pose_threshold:
      # Translation threshold in meters
      translation: 50.0
      # Rotation threshold in radians
      rotation: 3.15
    init_hack:
      translation:
        x: 0.0
        y: 0.0
        z: 0.0
      quaternion:
        x: 0.0
        y: 0.0
        z: 0.0
        w: 1.0
      enabled: true
    # localizer specific configuration
    localizer:
      #underlying voxel grid configuration of the map
      map:
        capacity: 1000000
        min_point:
          x: -1000.0
          y: -1000.0
          z: -3.0
        max_point:
          x: 1000.0
          y: 1000.0
          z: 3.0
        voxel_size:
          x: 3.5
          y: 3.5
          z: 3.5
      # ndt scan representation config
      scan:
        capacity: 100000
      # ndt optimization problem configuration
      optimization:
        outlier_ratio: 0.55 # default value from PCL
      # newton optimizer configuration
      optimizer:
        max_iterations: 40
        score_tolerance: 0.00001
        parameter_tolerance: 0.00001
        gradient_tolerance: 0.00002
        line_search:
          step_max: 0.12
          step_min: 0.0001
        # Maximum accepted duration between a scan and an initial pose guess
      guess_time_tolerance_ms: 750
